### DISCUSSION TITLE
Edge Impulse & Recording interaction

### DISCUSSION THREAD

**Discussion by OOHehir** [LIFsCode/ELOC-3.0#27] [General]
@LIFsCode It looks like the Edge Impulse (EI) is coming along, so far:
1. A per-recorded sound sample can be copied into a buffer & inference run on it successfully. This tests the EI code setup without an microphone interaction.
2. The I2S mic can be activated & data read from it & inference run on. This is buggy at the moment, could be due to power setting (light sleep) or a mic interaction issue.

With regards to moving forward at this point I'd like to raise a few things:
1. Power - I realize this is an important issue so whats the ideal setup? A noise wakes the ESP32 & then it starts running inference for say, the next 30 sec & if nothing is heard goes back to sleep? Is this possible with the I2S mic? 
2. Bluetooth - is this normally active? Not sure if there's any point (in regular use) running inference when the Bluetooth is in use.
3. Recording sound - I understand that if a trigger sound (e.g. trumpet) is detected then the sound should be recorded for 30 sec (or similar) but what is the most useful in the following scenario:
    - t = 0   Device switches on
    - t = 23 Device starts sampling a sound that will eventually (i.e. 1 or 2 sec later) be determined to be a trumpet
    - t = 25 Device reports a trumpet sound
    - t = 32 Device reports a second trumpet
    - t = 59 Device reports a third trumpet
   If this scenario should the device start recording at t = 23 (i.e. require continuous buffering of sound to be saved) or t = 25?
   While recording is in progress is there any point continuing to run inference?
   Should  the device restart the 30 sec recording time at t = 32 (assuming inference still running)?
   Should  the device restart the 30 sec recording time at t = 59 (assuming inference still running)?
   The possible issue here is that the SD could be filled quickly if a few elephants decide to 'hang out' nearby for a few hours..

Please let me know what you think & if there's any other possible issues..

Regards,

Owen

**Comment by LIFsCode** [comment]
@OOHehir can you be more specific what's the problem with the mic?
Did you try any other Software to check if the mic is working?
Do you still have the original HW or did you get a new HW in the meantime? if yes do both HW show the same effects?
@EDsteve : Do you have the same issues with the mic?

Concerning the other stuff:

1. Power: No light sleep with I2S mic is not possible, as light sleep will stop the internal clocks which are needed for I2S
2. Bluetooth: In my opinion BT should be disabled during recording. It can be reenabled via the Lis3DH tap sensor if needed. I would only keep BT on if we fully integrated BT LE, but this would require an external oscillator for the RTC (https://github.com/LIFsCode/ELOC-3.0/issues/11)
3. Recording Sound: @EDsteve  that's up to you

regards
Fabian

**Comment by OOHehir** [comment]
@LIFsCode @EDsteve Apologies for the misunderstanding, the issue with the mic is all my code, I'm trying to integrate the EI framework & it'll just take a bit more work, I don't think there's any big concern about it.
1. Thanks for the info
2. As is does the BT start up & then go to an 'idle'/ sleep state?


Thanks

**Comment by EDsteve** [comment]
@OOHehir @LIFsCode 
Great to read you made progress.

1. For now the AI should always run. The wake up functionality would reduce power consumption by a lot but needs to be done with an analog mic and some sort of filter. But this will be implemented later on. 

2. Yes. BT starts when the device is starting and after 60 seconds (i think) BT turns off and can be reactivated with tapping on the ELOC. Sound activation would be a very good solution as well. 

3. I just had a talk with Alex about it and he suggest to have 24/7 recording including inference during the fiirst months of testing. IF recording and inference is easily possible at the same time. 
If the 30 second solution is easier. This worrks as well.
How long the recording should be or if the recording should start again when an elephant sound is triggered within the 30 second time window depends on how complex it is to implement that. Both would work for us. **Most important is that we have the recognized sound on the SD card so we can check later on of it really was an elephant sound.**
That means it should start recording at t=23 but only when an elephant sound was detected :) Is that possible without too much effort? 

Main thing is to get a device ready for testing. Fine tuning can be done later. So whatever makes it easier to implement :)
All recorders will have an SD card with 128GB or so which is big enough for a few weeks of 16K recording. 

Cheers
ED

**Comment by OOHehir** [comment]
@EDsteve @LIFsCode 

1. I think I've isolated the issue with the I2S mic, it was probably not setting the correct config (getI2sConfig()) before starting it. It occasionally fails to register correctly at startup so I'll add some code to attempt resets of it (the I2SMEMSSampler). It might be as I'm running the board just on a USB to serial adapter & getting occasional brownouts/ low voltage drops.
2. _24/7 recording including inference_ Do you mean:
      A. Continuously record 24/7, regardless of whether an elephant is detected or
      B. Continuously run inference 24/7 & record **when** an elephant is detected?
  Both should be possible but can't be sure at this point, it probably depends on how CPU intensive writing to the SD card is.
  If necessary starting at t=23 shouldn't be much of an issue either.
  Perhaps it's also (when this is working) writing to a csv file a timestamp & inference predictions so the sound & results can be cross referenced for analyses?

Regards,

Owen

**Comment by EDsteve** [comment]
@OOHehir 
1. Sounds like you got it under control and you can do inference using the microphone now?
From my quick Goole search i read that some USB to TTL adapter only deliver 50mA at 3,3V. That's not enough.  Are you working with the ELOC 3.2 PCB? Switch the Battery switch to LiIon and you can connect anything from 3-6V to the battery plug.
2. Number A is what the biologists want. But if it makes things more complicated. The other solutions will do as well. 

Good point: Having a CSV file with all the time stamps is something very useful :)

Cheers
ED

**Comment by OOHehir** [comment]
@EDsteve Well, I can run the inference in a loop but the device is either suffering brownout or the results are not changing from one inference to the next, despite a change in sound. Quite likely be a power supply issue. 

Can the device be powered via the USB C port while running a serial connection?

I'll aim to implement solution A & let you know how I get on.

@EDsteve @LIFsCode 

As discussed before a simple model of the sound input & analysis is:

| <----------------------- A ---------------------->| <---------------------- B --------------------------------------->|
 sound ----> mic ---> I2S ---> DMA ---> Buffer --> Preprocessing --> Tensorflow model ---> Result

I've implemented code to test stage B but testing stage A is trickier. It would be useful to implement some sort of test to ensure stage A is performing correctly, & if not at which point the issue is. Some solutions I've considered:

1. Playing a sound (from another device) & observing the results on the device & comparing this against the same sound tested against a desktop version of EI. Not a great solution. 
2. Bypassing the mic & injecting appropriate signals from a function generator (or similar) to replicate the mic output. By injecting a known signal the result can be analyzed, a better solution. 
3. Connecting the I2S output of the ESP32 to its own I2S input, a 'loopback' scenario. Or if necessary using another ESP32 to supply the I2S output. Probably the best solution (if it works) as could be implemented as autonomous test code. 

I have a ESP-PROG on the way to help with debugging, it'll hopefully allow inspection of the buffers etc. to track down any issues. If you have any thoughts on the above they'd be most welcome.

Regards,

**Comment by OOHehir** [comment]
Seems [loopback](https://docs.espressif.com/projects/esp-idf/en/v5.1.1/esp32/api-reference/peripherals/i2s.html#introduction) is possible:

_DIN/DOUT: Serial data input/output line. Data will loopback internally if DIN and DOUT are set to a same GPIO._

**Comment by EDsteve** [comment]
@OOHehir 
The ELOC should run from USB C (**Turn the battery switch to LiIon**). But USB might add a lot of noise which might interfere with the recordings. Using a battery eliminates that issue.
If i remember correctly our firmware checks the voltage after boot and if the battery voltage is under 2.9V or so (or if no battery is connected, this voltage is 0), The ELOC goes into sleep. You might need to remove the voltage check for that purpose in our firmware.

About the I2S testing:
Are you testing stage A with our firmware or the EI firmware/"I2S-driver"?
If i understand you correctly you want to find out if stage A is working correctly. Maybe i missunderstand, but testing stage A should work by just recording the buffer to an SD card and listen to the recording? Or not accurate enough? :)

ESP-PROG seems like a good weapon against problems :)

I will be on my way back to Sumatra tomorrow. Testing and answering mails might be slower but will be done :)

Cheers
ED

**Comment by OOHehir** [comment]
@EDsteve 
Thanks, I've powered the board via the battery terminal & seems to have eliminated the brownout issues. The issue now is that the EI predictions aren't changing so the issue is in stage A, but hard to work out exactly where. I'm trying to adapt the ELOC firmware to incorporate the EI model, will continue working on it!

**Comment by OOHehir** [comment]
@EDsteve Well I think I worked out why the predictions weren't changing..
I went back & created a fresh project with the master branch (i.e without EI). The serial output looks normal (no errors) & a file is recorded on the SD card but its just silent. I tried powering via a serial adapter & an external power supply but no change. I can't recall if I ever tested the sound recording before on the ELOC board,  i think I just got stuck into the EI code. It seem like it might be a damaged mic. I've been fairly careful with the board so I don't think its damage I've caused, perhaps in transit?

**Comment by EDsteve** [comment]
@OOHehir 
That sounds like the common L/R problem. On the ELOC the L/R is connected to VCC.
Try to change to the other channel in SW (config.cpp?). If I remember correctly it should be the left channel.
I2S_CHANNEL_FMT_ONLY_LEFT

If the sound is completely silent (no noise) there are only three options in my experience:
1. Wiring wrong
2. Bad solder connection
3. L/R wrong, which also counts as wrong wiring in a way :)
4. If the battery switch was set to LiFePo4 and an input voltage of more than 3,6V was connected. Might kill the ADC in the microphone 

I am sure you will get it running. Let me know if that helped. I have tested the PCB before I shiped it. But if it really is damaged (my first time soldering with hot air solder). I can always send you another one.

**Comment by OOHehir** [comment]
@EDsteve Many thanks, as you suggested, changing the channel fixed that particular issue. The predictions are now changing in response to different sounds.

**Comment by LIFsCode** [comment]
@OOHehir which HW are you using? Just to avoid we have a wiring/config problem on the ELOC HW. I thought the current config in the code did match the HW

@EDsteve I have noticed some rare brownout conditions when running from USB. Most likely because BT may draw peek currents >600 mA during initialization.

@OOHehir concerning the inference do you have any data about the runtime? Just to get a hint of what performance is required and what performance is left.
I actually didn't really get your idea of data buffer and thread architecture in your code, so maybe we can align each other on the general concept.

**Comment by OOHehir** [comment]
@LIFsCode The new PCB (ELOC 3.2?). 

I had a look on a meter & think I've seen peaks of 900mA on startup, then settles down.

I attach a sample output of the board running with PERF_MONITOR output. If you need further details let me know.

Are you referring to how the continuous inference runs? There's an [explanation here](https://docs.edgeimpulse.com/docs/tutorials/advanced-inferencing/continuous-audio-sampling) which is probably better than I can explain it!

[sample_output.txt](https://github.com/LIFsCode/ELOC-3.0/files/12610251/sample_output.txt)

**Reply by EDsteve** [reply]
Yes. Owen has the ELOC 3.2

**Reply by LIFsCode** [reply]
@EDsteve is the mic working with yours? It seems strange if identical HW wiring with the same config ends up with different working states. Maybe a different series of the Mic IC has been assembled on one of the boards?

**Comment by EDsteve** [comment]
@OOHehir 
Happy to hear that the inference is working. Have you tried with a good model? How accurate is the prediction? If there is something for me to test. I will be happy to do so :)

@LIFsCode 
**Power Consumption**
I didnt expect the 900mA. But i guess if all adds up, 900mA seems plausible.
The LDO has a max output current of 1000mA. So i will change it to a bigger one with the neext HW revision. LoRa and GPS need some more juice too.
Using USB-C to power the board goes directly into the charging chip and from there to the power line. As far as i remember the battery charging chip has a max charging current of 800mA. So that might be the reason for the brown outs whoch shouldnt happen with a battery connected.

**Reply by OOHehir** [reply]
@EDsteve I've only used the model from your Rumble_Detector code so i don't think its very accurate. As soon as we develop an improved EI model it should be straightforward enough to modify.

**Comment by OOHehir** [comment]
@LIFsCode @EDsteve While waiting for an improved model I'll move onto integrating the recording. I notice that the current recording is provided by the function

```void record(I2SSampler *input)```

A few questions regarding this:

1. Should this be used for the continuous recording? 
2. I see it uses a timestamp of sorts for the filename. Should this be retained? 
3. Is the max recording for each file 60 sec & should this be changed?
4. Would there be any value in moving this function into ~the I2SSampler~ another class/ file?

If there's anything else you think is worth bringing up, please let me know.

**Reply by OOHehir** [reply]
@EDsteve @LIFsCode Edited previous comment regarding moving the function.

**Comment by LIFsCode** [comment]
@OOHehir This is what I tried to ask in my [post](https://github.com/LIFsCode/ELOC-3.0/discussions/27#discussioncomment-7002050) concerning the buffer and thread architecture.

I had an architecture in mind, which has a single I2S Sampler which captures data in a shared data buffer, signals EI task and SD-Card storage task, so that both can do their job.
So far I didn't implement any of these because I think we should agree on an architecture first.

Something like this
![grafik](https://github.com/LIFsCode/ELOC-3.0/assets/109753539/d4899971-acef-4694-9709-6762e9d1b555)


Which leaves me with your questions:

1. I wouldn't use recording function with the void `record(I2SSampler *input)` as is the I2S sampler should not be part of the recording engine, as it must run independently for the EI part.
2. Yes as I understood @EDsteve  this is the intention
3. The duration is part of the config and can be changed via BT
4. Yes this defninetly makes sense, currently there is still a lot of messy code. I wouldn't move it to I2SSampler, but could be integrated or derived from WAVFileWriter

@EDsteve : I'm still wondering why and how Tom chose the block sizes for writing to SD card. When I took over the code, he wrote in blocks of 16000 samples (with a DMA buffer size of 1000). The value seems to specific for pure coninscidence, but does not make sense to me. It neither matches SD card sector sizes nor 2^x sizes, so by first impression it seems obstructive concerning performance. 
Do you have any idea what was the reason for this?

@OOHehir :

1. Maybe a simple diagram would help to get us all on the same understanding of the SW architecture
2. I would prefer moving the inference stuff to one/more separte files/folder to keep an better overview

**Comment by OOHehir** [comment]
@LIFsCode I agree, moving the inference stuff into separate files would be cleaner. There are lots of dependencies with the edge impulse (EI) code so might take a bit of time..

I think your diagram makes sense, one complicating factor is probably (as far as I recall) the EI code modifies the contents of the buffer so we'll probably end up requiring separate buffers for the recording & EI.

@EDsteve @LIFsCode I'm not aware of the history of the code but just a note that may impact future decisions, from what I've seen so far nearly all the EI models are created at 16000Hz.

**Reply by LIFsCode** [reply]
@OOHehir ok that makes sense if the EI is modifying its input buffer. But still I would suggest that the EI consumer task (or any apropriate wrapper) would copy the buffer for the EI specific task. I would prefer that the responsibility for doubling/copying of buffers should always be in the hands of the tasks which require it, in this case the EI part.

This could keep the I2S Sampler simple.

Concerning architecture of the I2S Sampler & recording, & EI:
Keep in mind that the EI part will for sure increase the overall power consumption as it will add a significant amount of CPU time thus keeping the CPU out of idle and probably at higher clock rates.
Thus I would prefer having the I2S Sampler independent of EI logic, so that it is possible to enable/disable things separately:

- run everything
- disable recording, i.e. run in EI detection only mode, or only record after detectoin
- isable EI : i.e. run in recording only mode, with minimal power consumption, but without any detection logic.

@EDsteve : I don't know if this makes sense, you're the expert concerning the application scenarios

**Reply by EDsteve** [reply]
Yes. The ELOC should be able to switch between these three scenarios. The last two scenarios are mandatory. The first one (run verything) iss wished for if easy enough to impement :)

**Comment by EDsteve** [comment]
@LIFsCode
If the block sizes are the same as in the idf-wav-sd code, that means Tom did not change it. I will ask Tom about the block sizes, but the chances are high that he can't remember the reason any more. On the ogther hand I can do tests with different sizes if i am able to change the block sizes. Maybe it has an influence on power consumtption.

@OOHehir
All our dataset is based on 16K recordings. But when i train models with EI it seems a lower sample rate (e.g. 4K) works better for the. We just can't decide yet which sample rate will be used for the model.
But for the recording part, the sample rate is set in the app from the user. Maybe a good idea to install [our app](https://play.google.com/store/apps/details?id=de.eloc.eloc_control_panel) and play around with it. So you get a feeling of what is already there and how it all works.

I don't mind two different buffers as long we are able to write also the EI buffer to the SD card. What i mean is that there must be a way to check/see that the EI model gets the correct data fed. Hope that make sense.

**Comment by OOHehir** [comment]
@EDsteve @LIFsCode I've added EI 'log' file functionality to the ai branch. I attach a sample file:

[EI-results-ID-186372-DEPLOY-VER-2.csv](https://github.com/LIFsCode/ELOC-3.0/files/12795465/EI-results-ID-186372-DEPLOY-VER-2.csv)

This file is stored in the eloc directory & the file name is formatted as:

EI-results-ID-(edge impulse Project ID)--DEPLOY-VER-(edge impulse deploy version).csv

The contents are:
```
Year-Month-Day Hour:Min:Sec ,other ,trumpet
2023-10-03 17:23:19 , 0.94, 0.06
2023-10-03 17:23:23 , 0.99, 0.01
2023-10-03 17:23:28 , 0.98, 0.02
2023-10-03 17:23:32 , 0.98, 0.02
2023-10-03 17:23:37 , 1.00, 0.00
2023-10-03 17:23:41 , 1.00, 0.00
2023-10-03 17:23:46 , 1.00, 0.00
2023-10-03 17:23:50 , 1.00, 0.00
```
The last two digits give the probability of the respective sound (to two decimal places). The file is amended to,  not overwritten. If the file is not found (e.g. change of model) a new file is created with the respective name.

This is not fully tested but seems to be working without issue so far.

I'll move onto the sound recording..

**Reply by EDsteve** [reply]
Great. That will be very usefull. Thanks

**Comment by OOHehir** [comment]
@EDsteve @LIFsCode I've managed to make some progress on the sound processing, wav file creation & sound buffers. There's plenty still to improve but the highlights are as follows: 

**WAVFileWriter**
- This now creates two buffers in PSRAM that are multiples of 512 bytes (for optimum SD writing performance).  
- The length (in time) of the buffers is determined at instantiation. This allows for the buffers to continuously store sound & when a file recording is triggered to save the last X seconds of sound. It's currently set to 2 secs. 
- The class now incorporates a thread method that once started will write out to the wav file continuously once a buffer is filled (discussed below). 
-  Once the required file length is reached (determined by the config value secondsPerFile) the file is closed appropriately.
- Incorporates a 'Mode': 
  - enum class Mode { disabled = 0, single = 1, continuous = 2 }; 
  - Disabled - default at startup, recording not enabled
  - Single - a single file is recorded (length of secondsPerFile) once the AI threshold value is triggered (more discussion on this below)
  - Continuous - sound continuously saved in lengths of secondsPerFile.

**I2SEMSSampler**

- This incorporates a modified scheme as [proposed](https://github.com/LIFsCode/ELOC-3.0/discussions/27#discussioncomment-7035230).
- After instantiation the buffers of the edge impulse (EI/ AI) & wav writer are registered against the object. 
- Uses a temporary buffer in PSRAM to accept the incoming 32 bit samples from the I2S mic.
- Each sample is then converted to 16 bit & the volume level adjusted. The sample is then stored directly into both edge impulse buffer & the wav writer buffers. 
- The code is designed to automatically handle any difference in the I2S sample rate & the rate as required by the EI model, even if the EI model is changed. The limitation is that it should be an even divisor. I.e. if the I2S sample rate is 16kHz the EI sample rate should be 16kHz, 8kHz, 4kHz or 2kHz. 
- Once either buffer is detected as being full the appropriate buffer flag is set. 
- Has a mechanism to detect buffer overrun (mainly for debug purposes) 

**Main.cpp**  

I've modified significant portions of this, which hopefully won't case too many headaches for others but IMHO improve clarity including:
- Renamed variables, pointers etc to clarify their purpose
- Removed the record() function, implemented in the wav writer class
- Tried to improve commenting
- Improved some terminal output

If the changes don't make sense or you'd rather revert to previous no problem, please let me know.

The Edge Impulse code is still work-in-progress but a significant amount of the initial code is now handled by other classes so I'll remove it shortly. I'll continue working on moving it to its own class following help from @LIFsCode . 

**Suggestions**
I've seen some discussion on the Bluetooth JSON configurations & suggest that the following may wish to be added so that the can be modified by the user:
- AI threshold value - a float, (value < 1) at which a 'key sound' is determined to have been detected. Currently set [here](https://github.com/LIFsCode/ELOC-3.0/blob/41c1e6e5eb022373a51c35e6c942983411ef8afd/eloc610LowPowerPartition/src/config.h#L21). 
- Set the wav writer mode as outlined above. Currently set arbitrary [here](https://github.com/LIFsCode/ELOC-3.0/blob/41c1e6e5eb022373a51c35e6c942983411ef8afd/eloc610LowPowerPartition/src/main.cpp#L1282).
- Set the mode of the AI inference. Currently set [here](https://github.com/LIFsCode/ELOC-3.0/blob/41c1e6e5eb022373a51c35e6c942983411ef8afd/eloc610LowPowerPartition/src/main.cpp#L814). 
- I2S volume adjustment. This is currently set to '8' by I2S_VOLUME_SCALING_FACTOR [here](https://github.com/LIFsCode/ELOC-3.0/blob/41c1e6e5eb022373a51c35e6c942983411ef8afd/eloc610LowPowerPartition/lib/audio_input/src/I2SMEMSSampler.cpp#L109).

**Observations**
I realize that there is still a significant amount of code 'churn' on the AI branch given the development above but the longer we have two branches undergoing significant development the more likely they'll diverge. I'll try to update the AI branch shortly with changes from the master to try & keep it aligned.

I welcome any observations, criticism, thoughts on the above!

**Comment by LIFsCode** [comment]
@OOHehir great work ! The code looks much more structured now. I like the concept of having registered consumeres I2SMEMSSampler.

I did not have time to have a deeper look into your code, but from a first glance it looks really good. Thanks a lot.

Concerning the BT Interface I suggest the following:
### Control

Based on the current state of the [interface](https://github.com/LIFsCode/ELOC-3.0/wiki/ELOC-3.0-App-Interface-(proposal)) I would keep the control of AI Detection & Recoring in separate commands (Actually I'm not sure about that, there are also enough reasons to combine it as they depend on each other, so let's discuss this):

Just as a starting point for a discussion an idea for a concrete command spec:
| Command | Parameters | Description | Response |
| ------------- | ------------- | ------------- | ------------- |
| setRecordMode| \<mode\> <br> "off" <br> "on" <br> "singleTrig" <br> "continousTrig"| Enable/disable recording. If used without arguments, current mode is toggled(on/off). Otherwise set recording to specified mode, e.g. setRecordMode#mode=on<br> "off": no recording <br> "on": continous recording (independent of AI) <br> "singleTrig": **TBD**  <br> "continousTrig": **TBD** | JSON |
| setDetectMode| \<mode\> <br> "off" <br> "on"| Enable/disable AI detection. If used without arguments, current mode is toggled(on/off). Otherwise set detection to specified mode, e.g. setDetectMode#mode=on <br> "off": AI Detection disabled<br> "on": AI Detection running | JSON |

### Config Parameters
For this it is most easy to include it in the global JSON config as separate nested object

Suggestion could be:
```
{
  "AIdetection": {
    "threshold": 0.8,
    "volume": 8
  }
}
```

in combined view, e.g.

```
{
  "device": {
    "fileHeader": "not_set",
    "locationCode": "unknown",
    "locationAccuracy": 99,
    "nodeName": "ELOC_NONAME"
  },
  "config": {
    "secondsPerFile": 60,
    "cpuMaxFrequencyMHZ": 80,
    "cpuMinFrequencyMHZ": 10,
    "cpuEnableLightSleep": true,
    "bluetoothEnableAtStart": true,
    "bluetoothEnableOnTapping": true,
    "bluetoothEnableDuringRecord": true,
    "bluetoothOffTimeoutSeconds": 60
  },
  "mic": {
    "MicType": "ns",
    "MicBitShift": 11,
    "MicSampleRate": 20000,
    "MicUseAPLL": true,
    "MicUseTimingFix": true,
    "MicPointingDirectionDegrees": "ns"
  },
  "AIdetection": {
    "threshold": 0.8,
    "volume": 8
  }
}
```

> Set the wav writer mode as outlined above. Currently set arbitrary [here](https://github.com/LIFsCode/ELOC-3.0/blob/41c1e6e5eb022373a51c35e6c942983411ef8afd/eloc610LowPowerPartition/src/main.cpp#L1282).

> Set the mode of the AI inference. Currently set [here](https://github.com/LIFsCode/ELOC-3.0/blob/41c1e6e5eb022373a51c35e6c942983411ef8afd/eloc610LowPowerPartition/src/main.cpp#L814).

I'm not sure if this is covert by setting the setRecordMode/setDetectMode or if this sould be part of the config, I'm currently missing a bit of use case knowledge and a bit of understanding of the behavior of the code, so I just want to start a discussion.

### Config Parameters
Same here I would add the AI specific status as separate nested object to status JSON

@OOHehir @EDsteve : Any interesting status flags for the AI? I could image something like this:
```
{
  "AIdetection": {
    "EventsDetected": 42,
    "lastEventTime": 132456,
    "TimingAvg" : {
      "dsp[ms]" : 5,
      "classification[ms]" : 5,
      "anomaly[ms]" : 5
    }
  }
}
```
Also without much thought, just as a discusison base.

**Reply by OOHehir** [reply]
@LIFsCode Thanks, glad to be able to help!

Just some observations, suggestions:

- Regarding setRecordMode, I can only envisage three modes so far (off, continuous & triggered by AI detection) but perhaps @EDsteve has more use cases?
- Is a possible method of changing setRecordMode & setDetectMode at runtime by polling the config structure? Iḿ unsure what happens when values are changed from the Bluetooth side
- Regarding the volume I have to confirm by experiment (I'll expand further in the more relevant [discussion](https://github.com/LIFsCode/ELOC-3.0/discussions/24)) but can I suggest that the values be constrained to values of 1, 2, 4, 8, 16? The reason for this is that the 22 bit sound samples arrive from the DMA with the most significant byte at byte position 32 & need to be righted shifted by 8 bits to correct. Adjusting the volume in the above steps means that all that is required is to reduce the right shifting appropriately (i.e. double the volume by right shifting the volume 7 places, quadruple the volume by right shifting 6 places etc..). I have to confirm that this is correct but given that there is 16000 samples per second the above would keep processing time down significantly without (I hope) too much downside. 
- In conjunction with the above I think "MicBitShift" may be somewhat redundant. I would like to suggest that this value is moved to config.h (or similar) & that when compiling the firmware a board definition is created (e.g #define BOARD ELOC32). Based on the BOARD definition the relevant values would be set. This could be expanded to include ports/pins of peripherals, enable/disable features etc. This should simplify things when the current board is updated & the need exists to continue to support it.
- I imagine it would be useful to see realtime results via Bluetooth but @EDsteve will know better.

Regards,

**Reply by EDsteve** [reply]
@OOHehir @LIFsCode 
Great progress is happening :)

**setRcordMode**
continousTrig: I am guessing it means recording should start and continue "forever" when an elephant sound was detected.  Just had a quick chat about the modes with our Bilogists and that seems not necessary. We only need:
1. Detection OFF - Recording ON (Continues 24/7 recording, recording time per file can be set in app)
2. Detection ON - Recording once the AI triggers. Recording time is set in app (can be same time as recording time per file)
3. Detection ON - Recording ON (Continues 24/7 recording, recording time per file can be set in app)
4. Detection ON - Recording OFF

**AI status flags / realtime results**
To see the detected events in the app would be cool. Yes. :)
The name of the model should also be sent to the app.
